misk/src/main/kotlin/misk/web/actions/WebActionFactory.kt

```diff
@@ -76,35 +79,68 @@ internal class WebActionFactory @Inject constructor(
     val effectivePrefix = pathPrefix.dropLast(1)
 
     if (get != null) {
-      result += newBoundAction(provider, actionFunction,
+      collectBoundActions(result, provider, actionFunction,
           effectivePrefix + get.pathPattern, DispatchMechanism.GET)
     }
     if (post != null) {
-      result += newBoundAction(provider, actionFunction,
+      collectBoundActions(result, provider, actionFunction,
           effectivePrefix + post.pathPattern, DispatchMechanism.POST)
     }
     if (connectWebSocket != null) {
-      result += newBoundAction(provider, actionFunction,
+      collectBoundActions(result, provider, actionFunction,
           effectivePrefix + connectWebSocket.pathPattern, DispatchMechanism.WEBSOCKET)
     }
     if (grpc != null) {
-      result += newBoundAction(provider, actionFunction,
+      collectBoundActions(result, provider, actionFunction,
           effectivePrefix + grpc.path, DispatchMechanism.GRPC)
     }
 
     return result
   }
 
-  private fun <A : WebAction> newBoundAction(
+  private fun <A : WebAction> collectBoundActions(
+    result: MutableList<BoundAction<A>>,
     provider: Provider<A>,
     function: KFunction<*>,
     pathPattern: String,
     dispatchMechanism: DispatchMechanism
-  ): BoundAction<A> {
+  ) {
     // NB: The response media type may be omitted; in this case only generic return types (String,
     // ByteString, ResponseBody, etc) are supported
     val action = function.asAction(dispatchMechanism)
+    result += newBoundAction(provider, pathPattern, action)
+
+    // If we can create a synthetic action with a different media type, do it. This means all
+    // protobuf actions are also published as JSON actions.
+    val jsonVariant = transformActionIntoJson(action)
+    if (jsonVariant != null) {
+      result += newBoundAction(provider, pathPattern, jsonVariant)
+    }
+  }
+
+  /**
+   * Returns a copy of [action] that pretends JSON was the request and response content types.
+   * Returns null if neither of the input action's content types were protobuf.
+   */
+  private fun transformActionIntoJson(action: Action): Action? {
+    var jsonVariant = action
+
+    if (action.acceptedMediaRanges == protobufMediaRanges) {
+      jsonVariant = jsonVariant.copy(acceptedMediaRanges = jsonMediaRanges)
+    }
+    if (action.responseContentType == protobufMediaType) {
```

Comment:
what if only one of these is hit? Do we still want the json variant? I think yes, because the goal is to have a purely json request/response flavor